benchmarks/DAE/Crank.jmd

```diff
@@ -0,0 +1,584 @@
+end
+if KV != 0
+    GP[1,3+KV] = cosp2
+    GP[2,3+KV] = sinp2
+end
+
+# Forces - rigid motion entries.
+F = zeros(Num, 7)
+F[1] = -.5*l₁*γ*(m₁+2.0*m₂)*cosp1 +
+       -.5*l₁*l₂*m₂*V[2]*V[2]*sinp12
+F[2] = -.5*l₂*γ*m₂*cosp2 +
+        .5*l₁*l₂*m₂*V[1]*V[1]*sinp12
+F[3] = 0.0
+# Superposition of flexible motion (term f^e).
+F[1] += ρ*l₁*V[2]*V[2]*(-sinp12*c1TQ+cosp12*c2TQ) +
+        - 2.0*ρ*l₁*V[2]*(cosp12*c1TQD+sinp12*c2TQD)
+F[2] += ρ*l₁*V[1]*V[1]*(sinp12*c1TQ-cosp12*c2TQ) +
+        - 2.0*ρ*V[2]*c12TQD - 2.0*V[2]*QDTMQQ +
+        - ρ*QDTBQQD - ρ*γ*(cosp2*c1TQ-sinp2*c2TQ)
+# Coriolis and gravity terms flexible motion (Gamma).
+for I = 1:length(Q)
+    F[3+I] = V[2]*V[2]*MQQ[I] +
+           + ρ*(V[2]*V[2]*c12[I] +
+                l₁*V[1]*V[1]*(cosp12*c1[I]+sinp12*c2[I]) +
+                2.0*V[2]*BQQD[I]) +
+           - ρ*γ*(sinp2*c1[I]+cosp2*c2[I])
+end
+# Stiffness + damping terms - K q - D q`.
+for I = 1:length(Q)
+    F[3+I] = F[3+I] - KQQ[I] - DQQD[I]
+end
+
+#if IPAR[1] == 1
+#    # Nonlinear stiffness term
+#    FACK = 0.5*EE*BB*HH/l₂^2*π^2
+#    FACB = 80.0/(π^2*9.0)
+#    F[4] -= FACK*(Q[1]*Q[4]-FACB*Q[2]*(-4*Q[3]+2*Q[4]))
+#    F[5] -= FACK*(4*Q[2]*Q[4]-FACB*Q[1]*(-4*Q[3]+2*Q[4]))
+#    F[6] -= FACK*4.0*FACB*Q[1]*Q[2]
+#    F[7] -= FACK*(0.5*Q[1]^2+2*Q[2]^2-2*FACB*Q[1]*Q[2])
+#end
+
+# Dynamics part II ( M*w - f + G(T)*lambda ).
+DELTA = zeros(Num, 7)
+for I = 1:7
+    DELTA[I] = dot(AM[:,I], DDX) +
+               - F[I] + GP[1,I]*λ₁+GP[2,I]*λ₂+GP[3,I]*λ₃
+end
+
+# Acceleration level constraints.
+#ALC = zeros(Num, 3)
+#QDKU = KU > 0 ? QD[KU] : Num(0)
+#QDKV = KV > 0 ? QD[KV] : Num(0)
+#ALC = zeros(Num, 3)
+#ALC[1] = -l₁*sinp1*V[1]^2 - (l₂+QKU)*sinp2*V[2]^2 +
+#    2.0*V[2]*(cosp2*QDKU-sinp2*QDKV) - cosp2*V[2]^2*QKV
+#ALC[2] =  l₁*cosp1*V[1]^2 + (l₂+QKU)*cosp2*V[2]^2 +
+#    2.0*V[2]*(sinp2*QDKU+cosp2*QDKV) - sinp2*V[2]^2*QKV
+#ALC[3] = 0.0
+#for I = 1:7
+#    ALC[1] += GP[1,I]*DDX[I]
+#    ALC[2] += GP[2,I]*DDX[I]
+#    ALC[3] += GP[3,I]*DDX[I]
+#end
+
+## Velocity level constraints.
+#VLC = zeros(Num, 3)
+#VLC[1] = 0.0
+#VLC[2] = 0.0
+#VLC[3] = -Ω
+#for I = 1:7
+#    VLC[1] += GP[1,I]*DX[I]
+#    VLC[2] += GP[2,I]*DX[I]
+#    VLC[3] += GP[3,I]*DX[I]
+#end
+
+## Position level constraints.
+PLC = zeros(Num, 3)
+PLC[1] = l₁*sinp1 + l₂*sinp2 + QKU*sinp2 + QKV*cosp2
+PLC[2] = x₃ - l₁*cosp1 - l₂*cosp2 + -QKU*cosp2 + QKV*sinp2
+PLC[3] = ϕ₁ - Ω*t
+
+eqs = Symbolics.Equation[]
+for D in DELTA
+    push!(eqs, 0 ~ D)
+end
+for P in PLC
+    push!(eqs, 0 ~ P)
+end
+for I = 1:7
+    push!(eqs, D(X[I]) ~ DX[I])
+    push!(eqs, D(DX[I]) ~ DDX[I])
+end
+eqs
+```
+
+Create the DAE solver:
+
+```julia
+@named sys = ODESystem(eqs, t)
+simpsys = structural_simplify(sys)
+odaeprob = ODAEProblem(simpsys,[],(0.0,0.1))
+# mmprob = ODEProblem(sys,[],(0.0,0.1))
+daeprob = DAEProblem(sys,ivs,[],(0.0,0.1))
+
+probs = Any[nothing, nothing, nothing]
+refs = Any[nothing, nothing, nothing]
+# mm_sol = solve(mmprob,?,abstol=1/10^14,reltol=1/10^14);
```

Comment:
The lowering of higher derivatives to dummy values occurs in `structural_simplify`, so if you supply an equation with second derivatives you can only solve the simplified form. We should throw an error in this case because we know it's wrong, but 😅 https://github.com/SciML/ModelingToolkit.jl/issues/1389 we'll get to it haha.

Reply:
So `ODAEProblem would've been the only one that worked because in the original script that's the only one that used the simplified version of the system.

Reply:
It could not solve the simpsys either. I had to add dummy variables for both the first and second derivatives.

Reply:
Since the problem is given in index 2 form, it might need dummy derivatives for the simplification to give a stable solution, which is something the current simplification technique doesn't do. This might be a good case for @YingboMa to look at.